### PR TITLE
Make Variations for Payment Block Visible on Click

### LINKED ISSUE #28523: [Earn] Fix the Payments block selector
Originally spotted by @n3f:

> The payments block appears that lets you choose:

> But only the middle button works... Anyone else seeing something similar. (It's the same for me on trunk).

I was trying to point out when did this problem occur, and even after rolling back to mid-december, the problem is still there. It seems unrelated to our paid-newsletter work, but nevertheless - seems like a problem

## Reproduction path

This will be only possible on WPCOM testing flow as the premium content block is only available on wpcom:

1. Sync to wpcom
2. insert `/payment` on a site with enabled payment features
3. Observe that donation and premium content blocks are not able to be selected.

### PR DESCRIPTION
**Pull request by romarioraffington**
Fixes #28523

## Proposed changes:

There may be times when the inner blocks rendered by the Payment block are hidden. Update the visibility setting of the inner block on click if the block is hidden

https://user-images.githubusercontent.com/7076981/231888432-3414b821-8756-4ca1-9d8c-1306398153ff.mov

### Other information:

- [ ] Have you written new tests for your changes, if applicable?
- [x] Have you checked the E2E test CI results, and verified that your changes do not break them?
- [x] Have you tested your changes on WordPress.com, if applicable (if so, you'll see a generated comment below with a script to run)?

## Does this pull request change what data or activity we track or use?

No

## Testing instructions:

* Sync changes to your sandbox
* Disable the Donation block in Preferences > Blocks ( see video above )
* Insert the Payments block
* Select the Donations block
* Notice the Donations block is displayed

### CODE REVIEW

File: projects/plugins/jetpack/extensions/blocks/payments-intro/edit.js [hunk 1]
```diff
@@ -12,20 +12,37 @@ import defaultVariations from './variations';
...
 			blockType: getBlockType( name ),
 			hasInnerBlocks: getBlocks( clientId )?.length > 0,
 			hasPatterns: __experimentalGetAllowedPatterns?.().filter( patternFilter ).length > 0,
 		};
 	} );
 
-	const { replaceBlock, selectBlock } = useDispatch( blockEditorStore );
+	const { replaceBlock, selectBlock, updateSettings } = useDispatch( blockEditorStore );
+
+	const maybeUpdateBlockVisibility = useCallback(
```
Comment (millerf) [inline projects/plugins/jetpack/extensions/blocks/payments-intro/edit.js]: maybe the name should imply that the block gets visible? Like `maybeSetBlockVisible`?

**Review by millerf — COMMENTED** [review]
Code looks fine but the updateSettings does not update the global settings:


https://user-images.githubusercontent.com/790558/232094909-cc8aaa90-990b-401f-932a-1dfd0a8cf0f9.mov



.
.
.
 I tried to publish a post with a donation block, and the donate button was hidden. I feel like it might be related?

Screenshot 2023-04-14 at 17 51 21

**Comment by romarioraffington** [comment]
Hi Fab, thanks for taking a look.

> Code looks fine but the updateSettings does not update the global settings:

I actually think this is a good thing. I'm open to more discussion about this. 

Here's a possible use case:

1. User disables the Donation block
2. User inserts the Payment block
3. Donation block should be visible and work as expected since it's a part of the Payments block

Should we manually override the Donation block visibility for Gutenberg in this case or should we override the visibility only for the Donations block within the Payments block?


> I tried to publish a post with a donation block, and the donate button was hidden. I feel like it might be related?

I was unable to reproduce on my end.


https://user-images.githubusercontent.com/7076981/232131826-4b093b9c-c0f2-4dd0-b0d3-eabd7ef1ec21.mov


I took a look at your test site and it appears that the site would need to be upgraded in order to use the Donations block

Screenshot 2023-04-14 at 1 55 29 PM

File: projects/plugins/jetpack/extensions/blocks/payments-intro/edit.js (hunk 1 above)
Comment (romarioraffington) [reply projects/plugins/jetpack/extensions/blocks/payments-intro/edit.js]: Fixed in [5064fed](https://github.com/Automattic/jetpack/pull/30095/commits/5064fed9981dedcd09be49821486a13bd815788c)

**Comment by millerf** [comment]
@romarioraffington 

> Should we manually override the Donation block visibility for Gutenberg in this case or should we override the visibility only for the Donations block within the Payments block?
> 

I think you have a point. I forgot there was both of them.

> I was unable to reproduce on my end.

I think the issue was I didn't click on the $10 button. This UI is quite confusing, not counting for the template colour scheme.

> I took a look at your test site and it appears that the site would need to be upgraded in order to use the Donations block

That is because you are not using the test store. This site is not on a free plan as we are using Newsletters.

**Review by millerf — APPROVED** [review]
I am approving as I got confused on my previous review.
LGTM :shipit: once rebased and those pesky lint errors are gone...

**Comment by millerf** [comment]
@romarioraffington Regarding the settings not saving. I agree it is fine to not save them.

However, this seems to be due to [a bug reported here](https://github.com/WordPress/gutenberg/issues/15993). So once this is fixed, the setting will be updated.
So we need to somehow add the block and then revert.

Knowing that, I wonder if we should revert to greying/hiding the blocks that are disabled, despite being in the Payments block...

Let me know your thoughts (and please tag me in the response, so I get a ping)

**Review by millerf — COMMENTED** [review]
Reverting my approval after [some discoveries](https://github.com/Automattic/jetpack/pull/30095#issuecomment-1511205941)

**Comment by romarioraffington** [comment]
> Knowing that, I wonder if we should revert to greying/hiding the blocks that are disabled, despite being in the Payments block...
> Let me know your thoughts (and please tag me in the response, so I get a ping)

[Based on our user-facing documentation](https://wordpress.com/support/wordpress-editor/blocks/payments/#payments-options), we share that the Payments block contains the Donation, Payment Button and Premium blocks. I think greying / hiding these blocks would cause some confusion for the user and potentially lead to more support tickets.

From our docs:
> The Payments feature has multiple block options for various kinds of payments, including the [Donations Form](https://wordpress.com/support/wordpress-editor/blocks/donations/) block, the Payment Button block (this guide), and the [Premium Content](https://wordpress.com/support/wordpress-editor/blocks/premium-content-block/) block.

From [this comment](https://github.com/Automattic/jetpack/issues/28523#issuecomment-1511104797):

> An earlier investigation found that the root cause for the weird allowed blocks state is possibly due to p2: p1678770434676469/1675417267.339259-slack-C45SNKV4Z

[It appears that the block hidden issue](https://github.com/Automattic/p2editor/pull/586) may have been caused by P2. I'll [double-check that the P2 code is not causing this issue](https://github.com/Automattic/p2/issues/6103).

cc: @millerf

**Comment by romarioraffington** [comment]
>[It appears that the block hidden issue](https://github.com/Automattic/p2editor/pull/586) may have been caused by P2. I'll [double-check that the P2 code is not causing this issue](https://github.com/Automattic/p2/issues/6103).

Based on my testing the `enqueue_block_editor_assets` fn for P2s was not invoked.

See fbhepr%2Skers%2Sjcpbz%2Sjc%2Qpbagrag%2Scyhtvaf%2Sc2%2Sfreire%2Spber%2Srqvgbe%2Spynff%2Qrqvgbe%2Qpber.cuc%3Se%3Q726p4929%2346-og

**Review by jeherve — APPROVED** [review]

**Comment by millerf** [comment]
@romarioraffington Great work! And thanks for investigating about the issue with updateSettings. Asking async why the settings are not saved.
But it can be merged for now :shipit:

**Pull request merged**